cargo-clone-core/src/lib.rs

```diff
@@ -159,11 +160,19 @@ fn select_pkg<'a, T>(
 where
     T: Source + 'a,
 {
-    src.update()?;
+    src.invalidate_cache();
 
     let dep = Dependency::parse(name, vers, src.source_id())?;
     let mut summaries = vec![];
-    src.query(&dep, &mut |summary| summaries.push(summary))?;
+
+    loop {
+        match src.query(&dep, QueryKind::Exact, &mut |summary| {
+            summaries.push(summary)
+        })? {
+            std::task::Poll::Ready(()) => break,
+            std::task::Poll::Pending => src.block_until_ready()?,
+        }
+    }
```

Comment:
The `query` function returns a `Pool`.
[This](https://github.com/rust-lang/cargo/blob/d95124ee93d28bfe6960602130fd083701a3525d/src/cargo/ops/cargo_add/mod.rs#L525) is how cargo waits for the poll to be ready apparently.